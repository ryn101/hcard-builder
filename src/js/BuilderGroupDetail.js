import Unique from './Unique';

/**
 * Class for storing builder group detail specific information
 */
class BuilderGroupDetail extends Unique {
    constructor(description, modelName) {
        super('bgd');
        this.description = description;
        this.modelName = modelName;
    }
}

export default BuilderGroupDetail;
