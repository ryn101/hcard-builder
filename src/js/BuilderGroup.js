import Unique from './Unique';
import BuilderGroupDetail from './BuilderGroupDetail';

/**
 * Class for storing builder group specific information
 */
class BuilderGroup extends Unique {
    constructor(title, modelName = null, details = []) {
        super('bg');
        this.title = title;
        this.modelName = modelName;
        this.details = details;
    }

    addDetail(description, modelName) {
        this.details.push(new BuilderGroupDetail(description, modelName));
    }
}

export default BuilderGroup;
