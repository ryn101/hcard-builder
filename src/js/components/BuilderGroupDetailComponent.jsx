import React, { Component } from 'react';
import PropTypes from 'prop-types';
import BuilderGroupDetail from '../BuilderGroupDetail';
import BuilderGroup from '../BuilderGroup';

class BuilderGroupDetailComponent extends Component {
    onDetailUpdated = (event) => {
        const { builderGroup, builderGroupDetail, onDetailUpdated } = this.props;

        onDetailUpdated(builderGroup.modelName, builderGroupDetail.modelName, event.target.value);
    };

    render() {
        const {
            builderGroup: { modelName: groupModelName },
            builderGroupDetail: { description, modelName: detailModelName },
            modelValue,
        } = this.props;

        const prefixedModelName = `builder_${detailModelName}`;

        return (
            <li className="flex-column">
                <label htmlFor={prefixedModelName}>{description}</label>
                <input
                    id={prefixedModelName}
                    name={prefixedModelName}
                    onChange={this.onDetailUpdated}
                    value={modelValue ? modelValue[groupModelName][detailModelName] || '' : ''}
                    type="text"
                />
            </li>
        );
    }
}

BuilderGroupDetailComponent.propTypes = {
    builderGroup: PropTypes.instanceOf(BuilderGroup).isRequired,
    builderGroupDetail: PropTypes.instanceOf(BuilderGroupDetail).isRequired,
    onDetailUpdated: PropTypes.func,
    // eslint-disable-next-line react/forbid-prop-types
    modelValue: PropTypes.object,
};

BuilderGroupDetailComponent.defaultProps = {
    onDetailUpdated: () => {},
    modelValue: null,
};

export default BuilderGroupDetailComponent;
