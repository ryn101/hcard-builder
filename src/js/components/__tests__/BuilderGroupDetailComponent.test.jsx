import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import BuilderGroupDetailComponent from '../BuilderGroupDetailComponent';
import BuilderGroupDetail from '../../BuilderGroupDetail';
import BuilderGroup from '../../BuilderGroup';

const builderGroup = new BuilderGroup('Test Group');
const builderGroupDetail = new BuilderGroupDetail('Test Detail', 'testDetail');

configure({ adapter: new Adapter() });

it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
        <BuilderGroupDetailComponent
            builderGroupDetail={builderGroupDetail}
            builderGroup={builderGroup}
        />,
        div,
    );

    ReactDOM.unmountComponentAtNode(div);
});

describe('once the builder group detail has rendered', () => {
    let builderGroupDetailNode;

    beforeEach(() => {
        builderGroupDetailNode = shallow(
            <BuilderGroupDetailComponent
                builderGroupDetail={builderGroupDetail}
                builderGroup={builderGroup}
            />,
        );
    });

    it('renders with state correctly', () => {
        expect(builderGroupDetailNode).toMatchSnapshot();
    });

    it('renders the title correctly', () => {
        const label = builderGroupDetailNode.find('label');
        expect(label.text()).toEqual(builderGroupDetail.description);
        expect(label.prop('htmlFor')).toEqual(builderGroupDetail.modelName);
    });

    it('renders the input correctly', () => {
        const input = builderGroupDetailNode.find('input');
        expect(input.prop('name')).toEqual(builderGroupDetail.modelName);
    });
});
