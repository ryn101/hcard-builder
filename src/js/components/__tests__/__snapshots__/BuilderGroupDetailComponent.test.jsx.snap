// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`once the builder group detail has rendered renders with state correctly 1`] = `
<li
  className="flex-column"
>
  <label
    htmlFor="testDetail"
  >
    Test Detail
  </label>
  <input
    name="testDetail"
    onChange={[Function]}
    type="text"
    value=""
  />
</li>
`;
